lib/rucio/daemons/replicarecoverer/suspicious_replica_recoverer.py

```diff
@@ -0,0 +1,200 @@
+
+                for rse in surls_to_recover:
+                    logging.info('replica_recoverer[%i/%i]: ready to declare %i bad replica(s) on %s: %s.',
+                                 worker_number, total_workers, len(surls_to_recover[rse]), rse, str(surls_to_recover[rse]))
+                    if len(surls_to_recover[rse]) > max_replicas_per_rse:
+                        logging.warning('replica_recoverer[%i/%i]: encountered more than %i suspicious replicas (%s) on %s. Please investigate.', worker_number, total_workers, max_replicas_per_rse, str(len(surls_to_recover[rse])), rse)
+                    else:
+                        declare_bad_file_replicas(pfns=surls_to_recover[rse], reason='Suspicious. Automatic recovery.', issuer='root', status=BadFilesStatus.BAD, session=None)
+                        logging.info('replica_recoverer[%i/%i]: finished declaring bad replicas on %s.', worker_number, total_workers, rse)
+
+        except (DatabaseException, DatabaseError) as err:
+            if match('.*QueuePool.*', str(err.args[0])):
+                logging.warning(traceback.format_exc())
+                record_counter('replica.recoverer.exceptions.%s', err.__class__.__name__)
+            elif match('.*ORA-03135.*', str(err.args[0])):
+                logging.warning(traceback.format_exc())
+                record_counter('replica.recoverer.exceptions.%s', err.__class__.__name__)
+            else:
+                logging.critical(traceback.format_exc())
+                record_counter('replica.recoverer.exceptions.%s', err.__class__.__name__)
+        except Exception as err:
+            logging.critical(traceback.format_exc())
+            record_counter('replica.recoverer.exceptions.%s', err.__class__.__name__)
+        if once:
+            break
+
+    die(executable='rucio-replica-recoverer', hostname=socket.gethostname(), pid=os.getpid(), thread=threading.current_thread())
+    logging.info('replica_recoverer[%i/%i]: graceful stop done', worker_number, total_workers)
+
+
+def run(once=False, threads=1, younger_than=3, cnt_threshold=10, rse_like='MOCK', max_replicas_per_rse=100):
+
+    """
+    Starts up the Suspicious-Replica-Recoverer threads.
+    """
+    client_time, db_time = datetime.utcnow(), get_db_time()
+    max_offset = timedelta(hours=1, seconds=10)
+    if type(db_time) is datetime:
+        if db_time - client_time > max_offset or client_time - db_time > max_offset:
+            logging.critical('Offset between client and db time too big. Stopping Suspicious-Replica-Recoverer.')
+            return
+
+    sanity_check(executable='rucio-replica-recoverer', hostname=socket.gethostname())
+
+    if once:
+        declare_suspicious_replicas_bad(once, younger_than, cnt_threshold, rse_like, max_replicas_per_rse)
+    else:
+        logging.info('Suspicious file replicas recovery starting %i threads', threads)
+        threads = [threading.Thread(target=declare_suspicious_replicas_bad,
```

Comment:
I wonder if we should even allow multi-threaded mode for this? Since the `get_available_suspicious_replicas` method does not split the workload.

Reply:
I have seen daemons not being consistent in this, so I introduced the multi-threaded mode as I have seen it at most cases. I can remove it, just let me know.